src/interface/meshblock_data.cpp

```diff
@@ -72,124 +82,129 @@ void MeshBlockData<T>::Add(const std::string &label, const Metadata &metadata) {
+  SetBlockPointer(src);
+  resolved_packages_ = src.resolved_packages_;
+
+  auto add_var = [=, &flags](auto var) {
+    if (!flags.empty() && !var->metadata().AnyFlagsSet(flags)) {
+      return;
+    }
+
+    if (shallow_copy || var->IsSet(Metadata::OneCopy)) {
+      Add(var);
+    } else {
+      Add(var->AllocateCopy(false, pmy_block));
+    }
+  };
+
+  if (names.empty()) {
+    for (auto v : src.GetCellVariableVector()) {
+      add_var(v);
+    }
+    for (auto fv : src.GetFaceVector()) {
+      add_var(fv);
     }
-    auto sv = sparse_map.find(name);
-    if (sv != sparse_map.end()) {
-      if (found) {
-        std::stringstream msg;
-        msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-        PARTHENON_THROW(msg);
+  } else {
+    auto var_map = src.GetCellVariableMap();
+    auto face_map = src.GetFaceMap();
+
+    for (const auto &name : names) {
+      bool found = false;
+      auto v = var_map.find(name);
+      if (v != var_map.end()) {
+        found = true;
+        add_var(v->second);
       }
-      found = true;
-      std::shared_ptr<SparseVariable<T>> newvar;
-      if (sparse_ids.size() > 0) {
-        newvar = std::make_shared<SparseVariable<T>>(sv->second, sparse_ids);
-      } else {
-        newvar = sv->second;
+
+      auto fv = face_map.find(name);
+      if (fv != face_map.end()) {
+        PARTHENON_REQUIRE_THROWS(!found, "MeshBlockData::CopyFrom: Variable '" + name +
+                                             "' found more than once");
+        found = true;
+        add_var(fv->second);
       }
-      Add(newvar);
-    }
-    auto fv = face_map.find(name);
-    if (fv != face_map.end()) {
-      if (found) {
-        std::stringstream msg;
-        msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-        PARTHENON_THROW(msg);
+
+      if (!found && (resolved_packages_ != nullptr)) {
+        // check if this is a sparse base name, if so we get its pool of sparse_ids,
+        // otherwise we get an empty pool
+        const auto &sparse_pool = resolved_packages_->GetSparsePool(name);
+
+        // add all sparse ids of the pool
+        for (const auto iter : sparse_pool.pool()) {
+          // this variable must exist, if it doesn't something is very wrong
+          const auto &v = varMap_.at(MakeVarLabel(name, iter.first));
+          add_var(v);
+          found = true;
+        }
       }
-      found = true;
-      Add(fv->second);
-    }
-    if (!found) {
-      std::stringstream msg;
-      msg << "MeshBlockData: " << name << " found more than once!" << std::endl;
-      PARTHENON_THROW(msg);
+
+      PARTHENON_REQUIRE_THROWS(found, "MeshBlockData::CopyFrom: Variable '" + name +
+                                          "' not found");
     }
   }
 }
+
+// Constructor for getting sub-containers
+// the variables returned are all shallow copies of the src container.
+// Optionally extract only some of the sparse ids of src variable.
```

Comment:
Where is that option?

Reply:
Great catch! Looks like I copied the comment but forgot to implement the feature. Will add it.